### PR TITLE
Add support for Metadata extension validation

### LINKED ISSUE #105: Add metadata extension & extension validation to provider structure
Metadata specification should also be a static yaml in papiea. Provider custom portion should be named extension and should be provided as another object model yaml by the provider at registration time. So for example:

      "metadata" : {
        "kind" : "imaged_cluster",
        "spec_version" : 1,
        "uuid" : "116e4001-1a80-4a50-9711-5114914d1311",
        "created_at" : ISODate("2019-04-04T18:15:22.501Z"),
        "extension": {      // The following section needs to be configured when registering the provider
          "owner" : "user@domain.com",
          "tenant_uuid" : "1ff18510-8151-4131-1f15-14161e113131"
        }
      }

### PR DESCRIPTION
**Pull request by Emulebest**
resolves #105

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -3,27 +3,30 @@ import createAPIDocsRouter from "./api_docs/api_docs_routes";
 import ApiDocsGenerator from "./api_docs/api_docs_generator";
 import createProviderAPIRouter from "./provider/provider_routes";
 import { Provider_API_Impl } from "./provider/provider_api_impl";
-import { Provider_API } from "./provider/provider_api_interface";
 import { MongoConnection } from "./databases/mongo";
 import { createEntityRoutes } from "./entity/entity_routes";
 import { Entity_API_Impl, ProcedureInvocationError } from "./entity/entity_api_impl";
 import { ValidationError, Validator } from "./validator";
 import { EntityNotFoundError } from "./databases/utils/errors";
+import * as morgan from "morgan";
```
Comment (adolgarev) [inline src/main.ts]: move logging to separate PR

File: src/databases/spec_db_mongo.ts [hunk 2]
```diff
@@ -24,17 +25,19 @@ export class Spec_DB_Mongo implements Spec_DB {
...
 
     async update_spec(entity_metadata: core.Metadata, spec: core.Spec): Promise<[core.Metadata, core.Spec]> {
+        let additional_fields: any = {};
+        if (entity_metadata.extension !== undefined) {
+            additional_fields = encode({"metadata.extension": entity_metadata.extension});
+        }
+        additional_fields["metadata.created_at"] = new Date();
+        if (entity_metadata["deleted_at"] !== undefined) {
+            additional_fields["metadata.deleted_at"] = entity_metadata["deleted_at"]
+        }
```
Comment (adolgarev) [inline src/databases/spec_db_mongo.ts]: think additional_fields["metadata.spec_version"] should be set as well

File: src/entity/entity_api_impl.ts [hunk 3]
```diff
@@ -36,15 +36,25 @@ export class Entity_API_Impl implements Entity_API {
...
+
+    private find_kind(provider: Provider, kind_name: string): Kind {
         const found_kind: Kind | undefined = provider.kinds.find(elem => elem.name === kind_name);
         if (found_kind === undefined) {
             throw new Error(`Kind: ${kind_name} not found`);
         }
         return found_kind;
     }
 
+    private async get_provider_by_kind(kind_name: string): Promise<Provider> {
```
Comment (adolgarev) [inline src/entity/entity_api_impl.ts]: this function is not needed, it is called once, it just calls provider_api

File: src/provider_sdk/typescript_sdk.ts [hunk 4]
```diff
@@ -17,6 +17,7 @@ export class ProviderSdk implements ProviderImpl {
     _provider: Provider | null;
     papiea_url: string;
     papiea_port: number;
+    private _meta_ext: { [key: string]: string };
```
Comment (adolgarev) [inline src/provider_sdk/typescript_sdk.ts]: TypeScript has private modifier, no need in leading underscore in name

File: start.sh [hunk 5]
```diff
@@ -5,7 +5,6 @@ if [ -d "/node_modules" ] && [ ! -d "node_modules" ]; then
     mv /node_modules node_modules
 fi
 
-export DEBUG=express:*
```
Comment (adolgarev) [inline start.sh]: this is related to logging, move to other PR

File: src/databases/spec_db_mongo.ts (hunk 2 above)
Comment (shlomi-nx) [reply src/databases/spec_db_mongo.ts]: created/deleted_at as well as spec_version are core to papiea and should not be in "additional_fields". 

Additional_fields should only be fields that are custom to the provider and that are not looked at from papiea without any specific configuration (like we do in casbin)

**Comment by shlomi-nx** [comment]
@Emulebest lets resolve and merge.
